locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/JcloudsLocationSecurityGroupEditor.java

```diff
@@ -0,0 +1,303 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.location.jclouds.networking;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import org.apache.brooklyn.util.exceptions.Exceptions;
+import org.jclouds.aws.AWSResponseException;
+import org.jclouds.compute.ComputeService;
+import org.jclouds.compute.domain.SecurityGroup;
+import org.jclouds.compute.extensions.SecurityGroupExtension;
+import org.jclouds.domain.Location;
+import org.jclouds.net.domain.IpPermission;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+/**
+ * Utility for manipulating security groups.
+ *
+ * For convenience, constructors take an argument of a {@link ComputeService} and extract from it the
+ * information needed to edit groups. However, not all compute services support the security group extension,
+ * so if the editor is being created for a compute service where it is not known in advance then users
+ * should query {@link #hasServiceSupport()} before calling the methods to edit security groups.
+ */
+public class JcloudsLocationSecurityGroupEditor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(JcloudsLocationSecurityGroupEditor.class);
+    public static final java.lang.String JCLOUDS_PREFIX = "jclouds#";
+
+    private final Location location;
+    private final ComputeService computeService;
+    private final Optional<SecurityGroupExtension> securityApi;
+    private final String locationId;
+    private final Predicate<Exception> isExceptionRetryable;
+
+    /**
+     * Constructor for editor that will retry operations upon exceptions.
+     * @param location JClouds location where security groups will be managed.
+     * @param computeService The JClouds compute service instance.
+     * @param predicate A predicate indicating whether the customiser can retry a request (e.g. to add a security group
+     * or a rule) if the attempted operation throws a Throwable.
+     */
+    public JcloudsLocationSecurityGroupEditor(Location location, ComputeService computeService, Predicate predicate) {
+        this.location = location;
+        this.computeService = computeService;
+        this.locationId = this.computeService.getContext().unwrap().getId();
```

Comment:
The `computeService` is only used for logging, and for getting the initial securitygroupExension.

The `locationId` is only used for the `getLocationId()` method, and for logging.
The `locationId` is also a bit confusing in Brooklyn, given we have a location class in Brooklyn as well. Its value is compared against "aws-ec2" - perhaps we should call it the `providerId`?

I wonder if we should pass less in - and perhaps expect the caller to have done the `computeservice.getSecurityGroupExtension().get()` call for us.